### PR TITLE
Retry /publicRooms a few times to allow for propagation delay before a new room appears in the room directory

### PR DESCRIPTION
**Pull request by reivilibre**
This is needed in Synapse because the room directory is updated in the background. Part of https://github.com/matrix-org/synapse/issues/13161.
It's particularly worse when using workers, because the room directory is updated in a different process and CPU contention increases the likelihood of hitting the race.

### CODE REVIEW

File: internal/client/client.go
Comment (DMRobertson) [inline internal/client/client.go:468]: Two points of confusion:

1. I think this means you can't use `MustDoFuncUntil` to repeat a successful call until it fails. E.g. imagine a test where:
  - join a room
  - someone bans you
  - repeatedly try to change your name in the room until the ban takes place, at which point the name change fails.
2. I think this means that you can't use `MustDoFuncUntil` to repeat a failing call until it suceeds. E.g. imagine a test where:
   - fail to join a restricted room because you're not in the right space
   - join the space
   - repeatedly join the restricted room until it succeeds.

Though thinking about it some more, perhaps those ought be different functions e.g. some sort of `DoFuncUntil`.

File: internal/client/client.go
Comment (DMRobertson) [inline internal/client/client.go:475]: If the average race is resolved quicker than this, say 100ms, then this is a little pessimistic. Might exponential backoff (rather than linear) make more sense? (e.g. 10ms delay, quadrupling every retry)

**Review by DMRobertson — COMMENTED** [review]
Drive-by thoughts. No idea what a Waiter is (https://pkg.go.dev/github.com/aws/aws-sdk-go/private/waiter#Waiter?) but if it's more idiomatic then that might be the way to go.

File: internal/client/client.go
Comment (reivilibre) [reply internal/client/client.go:468]: That's fair; not sure what better name to put on it (I've tried to keep the names consistent with what's there; I'd personally say those names aren't very intuitive either but they have docstrings).
At the very least, describing your caveats in the docstring would probably help though.

File: internal/client/client.go
Comment (kegsay) [inline internal/client/client.go:466]: Please don't add new functions which are basically the same as other functions. Add a `RequestOpt`.

File: internal/client/client.go
Comment (kegsay) [inline internal/client/client.go:479]: Please don't conflate performing operations (e.g HTTP requests) with verifying the results of said operation (assertions). This ends up making inflexible code.

File: tests/csapi/apidoc_room_state_test.go
Comment (kegsay) [inline tests/csapi/apidoc_room_state_test.go:112]: You can't use `MatchResponse` in this way as it will fail the test if the response does not pass the given matchers.

**Review by kegsay — CHANGES_REQUESTED** [review]
What we want is something which will repeat the action until some condition is satisfied, at which point it will return the response. This response can then have assertions applied to it. Currently, this PR conflates the two: the presence of the room ID is both the condition to satisfy _and_ the assertion. This doesn't compose well because:
 - You cannot use assertion helpers (like matchers) or anything which will fail the test as part of your code to repeat until the condition is satisfied, as the first failure will fail the test.
 - You cannot say things like "do this request until you see this field, then give me the response so I can inspect the fields more deeply".

File: internal/client/client.go
Comment (kegsay) [reply internal/client/client.go]: This is too inflexible. Some endpoints may take longer than others to update, it needs to be configurable how long to keep trying for.

File: tests/csapi/apidoc_room_state_test.go
Comment (reivilibre) [reply tests/csapi/apidoc_room_state_test.go]: (NB Diff is outdated but basic structure remains the same, so will reply to comment)

I believe this is what we want, though (or at least I'm used to it from SyTest); if the response is ever malformed then we want to fail the test immediately without waiting for it to be correct.

Put another way: The retries are waiting for the expected data to appear in the response, but that doesn't mean that any responses before that are allowed to omit required fields.
As a result, you end up with 3 outcomes: success, temporary fail (i.e. retry) and permanent fail.

As for how these currently seem to map to code:
Success = `return true`
Temporary fail = `return false`
Permanent fail = `t.FailNow`/`t.Fatalf`

**Review by erikjohnston — APPROVED** [review]
Seems legit to me. Will defer to @kegsay that this is the Right Way

File: tests/csapi/apidoc_room_state_test.go
Comment (kegsay) [inline tests/csapi/apidoc_room_state_test.go:123]: This is a slight misuse of matchers, given you always return `nil` it will always succeed even if the room isn't found, and then you side-effect `foundRoom` to determine retry-ness. It's not something that the matchers were designed for, but in practice it seems to read pretty well.

**Review by kegsay — APPROVED** [review]
LGTM

File: tests/csapi/apidoc_room_state_test.go
Comment (reivilibre) [reply tests/csapi/apidoc_room_state_test.go:123]: You're right, though I will say 'don't blame me, it was there before!' :-P

I suspect this sort of pattern will crop up again, so if you get any better ideas then it may be worth putting them in

**Pull request merged**
